src/quantization/sq4_uniform_quantizer.h

```diff
@@ -0,0 +1,245 @@
+// limitations under the License.
+
+#pragma once
+#include <math.h>
+
+#include <limits>
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+#include "quantizer.h"
+#include "simd/sq4_uniform_simd.h"
+
+namespace vsag {
+
+typedef uint64_t norm_type;
+typedef float sum_type;
+
+const char* const OFFSET_KEY_CODE = "code";
+const char* const OFFSET_KEY_NORM = "norm";
+const char* const OFFSET_KEY_SUM = "sum";
+
+template <MetricType Metric = MetricType::METRIC_TYPE_L2SQR>
+class SQ4UniformQuantizer : public Quantizer<SQ4UniformQuantizer<Metric>> {
+public:
+    explicit SQ4UniformQuantizer(int dim);
+
+    bool
+    TrainImpl(const DataType* data, uint64_t count);
+
+    bool
+    EncodeOneImpl(const DataType* data, uint8_t* codes) const;
+
+    bool
+    EncodeBatchImpl(const DataType* data, uint8_t* codes, uint64_t count);
+
+    bool
+    DecodeOneImpl(const uint8_t* codes, DataType* data);
+
+    bool
+    DecodeBatchImpl(const uint8_t* codes, DataType* data, uint64_t count);
+
+    inline float
+    ComputeImpl(const uint8_t* codes1, const uint8_t* codes2) const;
+
+    inline void
+    ProcessQueryImpl(const DataType* query, Computer<SQ4UniformQuantizer>& computer) const;
+
+    inline void
+    ComputeDistImpl(Computer<SQ4UniformQuantizer>& computer,
+                    const uint8_t* codes,
+                    float* dists) const;
+
+private:
+    DataType lower_bound_;
+    DataType diff_;
+    std::unordered_map<std::string, uint64_t> offsets_;
+};
+
+template <MetricType Metric>
+SQ4UniformQuantizer<Metric>::SQ4UniformQuantizer(int dim)
+    : Quantizer<SQ4UniformQuantizer<Metric>>(dim) {
+    lower_bound_ = std::numeric_limits<DataType>::max();
+    diff_ = std::numeric_limits<DataType>::min();
+
+    this->code_size_ = 0;
+
+    offsets_[OFFSET_KEY_CODE] = this->code_size_;
+    this->code_size_ += (dim + 1) >> 1 << 1;
+
+    if (Metric == MetricType::METRIC_TYPE_L2SQR or Metric == MetricType::METRIC_TYPE_COSINE) {
+        offsets_[OFFSET_KEY_NORM] = this->code_size_;
+        this->code_size_ += sizeof(norm_type);  // norm of vector
+    }
+
+    if (Metric == MetricType::METRIC_TYPE_IP or Metric == MetricType::METRIC_TYPE_COSINE) {
+        offsets_[OFFSET_KEY_SUM] = this->code_size_;
+        this->code_size_ += sizeof(sum_type);  // sum of vector
+    }
+}
+
+template <MetricType Metric>
+bool
+SQ4UniformQuantizer<Metric>::TrainImpl(const DataType* data, uint64_t count) {
+    if (data == nullptr) {
+        return false;
+    }
+
+    for (uint32_t i = 0; i < count; i++) {
+        for (uint32_t d = 0; d < this->dim_; d++) {
+            auto val = data[i * this->dim_ + d];
+            if (val > diff_) {
+                diff_ = val;
+            }
+            if (val < lower_bound_) {
+                lower_bound_ = val;
+            }
+        }
+    }
+
+    for (uint32_t d = 0; d < this->dim_; d++) {
+        diff_ -= lower_bound_;
+    }
```

Comment:
Can rewrite to `diff_ -= lower_bound_ * this->dim_` ?

Reply:
fixed